### PR TITLE
Allow MIR borrowck to catch unused mutable locals

### LINKED ISSUE #47279: NLL suggests removing mutability from a variable, doing so causes a compiler error
UPDATE: Mentoring instructions [below](https://github.com/rust-lang/rust/issues/47279#issuecomment-369030830).

---

This code compiles, but with a warning:

```rust
#![feature(nll)]

#[derive(Debug)]
struct A {}

fn init_a() -> A {
    A {}
}

#[derive(Debug)]
struct B<'a> {
    ed: &'a mut A,
}

fn init_b<'a>(ed: &'a mut A) -> B<'a> {
    B { ed }
}

#[derive(Debug)]
struct C<'a> {
    pd: &'a mut B<'a>,
}

fn init_c<'a>(pd: &'a mut B<'a>) -> C<'a> {
    C { pd }
}

#[derive(Debug)]
struct D<'a> {
    sd: &'a mut C<'a>,
}

fn init_d<'a>(sd: &'a mut C<'a>) -> D<'a> {
    D { sd }
}

fn main() {
    let mut a = init_a();
    let mut b = init_b(&mut a);
    let mut c = init_c(&mut b);

    let d = init_d(&mut c);

    println!("{:?}", d)
}
```

```
warning: variable does not need to be mutable
  --> src/main.rs:40:9
   |
40 |     let mut c = init_c(&mut b);
   |         ---^^
   |         |
   |         help: remove this `mut`
   |
   = note: #[warn(unused_mut)] on by default
```

Removing the `mut` causes a compiler error:

```
error[E0596]: cannot borrow immutable item `c` as mutable
  --> src/main.rs:42:20
   |
42 |     let d = init_d(&mut c);
   |                    ^^^^^^ cannot borrow as mutable
```

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #47279.

r? @nikomatsakis

### CODE REVIEW

**Comment by nikomatsakis** [comment]
This looks about right. There may be some complications around closures:

```rust
let mut x = &mut vec![]; 
//  ^^^ not needed
let closure = || {
    x.push(22);
};
closure();
```

Also, we have to disable the [other lint, from old AST borrowck](https://github.com/rust-lang/rust/blob/master/src/librustc_borrowck/borrowck/unused.rs)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This looks great! My main concern is the tests, left some suggestions below.

**Review by nikomatsakis — APPROVED** [review]

File: src/test/compile-fail/lint-unused-mut-variables.rs [hunk 1]
```diff
@@ -8,6 +8,9 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// revisions: lexical nll
+#![cfg_attr(nll, feature(nll))]
```
Comment (nikomatsakis) [inline src/test/compile-fail/lint-unused-mut-variables.rs]: We're getting [travis errors](https://travis-ci.org/rust-lang/rust/builds/350678719#L2394). The problem is that we have to update the `//~ ERROR` declarations in this file to account for the revisions.

For each line:

```
foo //~ ERROR bar
```

we have to change it to

```
foo //[lexical]~ ERROR bar
 //[nll]~^ ERROR bar
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by KiChjang** [comment]
Ok, so I figured out why Travis CI is complaining. Given the following code:

```rust
fn main() {
    let mut a = 3;
}
```

The current implementation sees an assignment to `a` during initialization, which makes it erroneously believe that the value has been mutated. We need to make it so that it understands that initialization is not equivalent with mutation.

**Comment by KiChjang** [comment]
Ok, so the latest commit still doesn't fix everything yet. It's quite over-eager in its job and flags too much fn args as being unnecessarily mutable. It also doesn't catch the following case:

```rust
    fn mut_ref_arg(mut arg : &mut [u8]) -> &mut [u8] {
        &mut arg[..] //[lexical]~^ ERROR: variable does not need to be mutable
                     //[nll]~^ ERROR: variable does not need to be mutable
    }
```

**Comment by nikomatsakis** [comment]
> We need to make it so that it understands that initialization is not equivalent with mutation.

We should be able to use the initialization state for this. Presumably we are doing a similar analysis already to report errors in the case that the variable is *not* `mut`...

Ah, indeed we do, right here:

https://github.com/rust-lang/rust/blob/8c4ff22a2d745097197c659ef9e3b04b8ceeb070/src/librustc_mir/borrow_check/mod.rs#L1279-L1303

File: src/librustc_mir/borrow_check/mod.rs [hunk 2]
```diff
@@ -1547,6 +1614,23 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 err.emit();
             },
             Reservation(WriteKind::Mutate) | Write(WriteKind::Mutate) => {
+                match place {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: I think logic is too simplistic. For example, consider this case:

```rust
let mut x = (1, 2);
x.0 = 22;
```

Here, the mutated place will be a field projection, but we still need to record `x` as a "used mut". We want to integrate more with the `is_mutable` code, I think -- in particular, maybe it should return the "root reason" that a path was mutable, even on success (right now it does that only on error). Then, if that root reason is a Local, we can add it to the `used_mut` set? etc

**Comment by KiChjang** [comment]
I've just pushed a new commit. The testst still fail, but I think we've caught most of the false positives (aside from 2), but in exchange we have 4 false negatives outstanding.

First, the false positives:
```rust
let x = |mut y: isize| 10; // correctly flags mut y, but also contains a spurious one that spans the entire RHS of the statement

let x = |mut y: isize| y = 32; // similar to the above, flags entire RHS, but does not flag mut y
```

And the false negatives:
```rust
let mut a = &mut 5;

let mut a = 5;
let mut b = (&mut a,);

fn mut_ref_arg(mut arg : &mut [u8]) -> &mut [u8] {
    &mut arg[..]
}

let mut v : &mut Vec<()> = &mut vec![];
```

This looks very much like that we're being overzealous in marking mutable references as being used.

The entire MIR dump for `lint-unused-mut-variables.rs` is located here: https://gist.github.com/KiChjang/80b513016e546ef5f08c43313bec7686

**Comment by nikomatsakis** [comment]
@KiChjang no time for details, but that is an interesting case! The thing is that if you have an `&mut` reference, then it must be located in a *unique* (but not necessarily *mutable*) location in order to use it mutably. So e.g. if I have `let x: &mut u32`, I can do `*x += 1`, but if I have `let x: &&mut 32`, I cannot do `**x += 1`. This disconnect is "the other place" where we wind up checking for *uniqueness* (but not mutability) -- i.e., the other place that gives rise to that flag I was talking about. I'll try to look a bit more deeply later, but maybe that gives you a clue what needs to change, not sure.

File: src/librustc_mir/borrow_check/mod.rs [hunk 3]
```diff
@@ -1524,47 +1591,91 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
-                    Some(name) => format!("immutable item `{}`", name),
-                    None => "immutable item".to_owned(),
-                };
-
-                let mut err = self.tcx
-                    .cannot_borrow_path_as_mutable(span, &item_msg, Origin::Mir);
-                err.span_label(span, "cannot borrow as mutable");
+            | Write(WriteKind::MutableBorrow(BorrowKind::Mut { .. })) => {
+                match self.is_mutable(place, is_local_mutation_allowed) {
+                    Ok(Place::Local(local))
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: Nit: I find this setup confusing. Can we do this instead?

```rust
Ok(Place::Local(local)) => {
    if is_local_mutation_allowed != ... { ... }
}
```

and then remove the *other* `Place::Local` arm below? 

(In general, I prefer to avoid `if` clauses on match arms except in particular cases, e.g. with very trivial bodies or where it enables a lot more re-use.)

**Comment by nikomatsakis** [comment]
OK. I'm not sure the best way to *fix* the problem, but let me elaborate a bit more on what is going on, and propose one possible fix.

First off, observe that the following [is legal](https://play.rust-lang.org/?gist=9f662a47be41e9abf87437b76699b336&version=nightly):

```rust
#![feature(nll)]

fn main() {
    let mut a = 5;
    let b = &mut a;
    let c = &mut *b;
}
```

The key point here is that `b` need not be declared as `mut`, even though we are borrowing `*b` as mutable, because the mutability comes from the `&mut` type. *However,* this does not mean that you can just do a `mut` borrow of a `&mut` reference any old time -- it must be found in a "unique" path (like a local variable).

In the code, this is reflected by a mode switch in `is_mutable`. When we encounter a deref place:

https://github.com/rust-lang/rust/blob/75af15ee6ca0c12b699a17984b033363cd25e9c3/src/librustc_mir/borrow_check/mod.rs#L1623-L1625

If the path being dereferenced is some kind of reference:

https://github.com/rust-lang/rust/blob/75af15ee6ca0c12b699a17984b033363cd25e9c3/src/librustc_mir/borrow_check/mod.rs#L1629-L1630

and in particular a **mutable** reference:

https://github.com/rust-lang/rust/blob/75af15ee6ca0c12b699a17984b033363cd25e9c3/src/librustc_mir/borrow_check/mod.rs#L1636

then we compute a new `mode` for local variables:

https://github.com/rust-lang/rust/blob/75af15ee6ca0c12b699a17984b033363cd25e9c3/src/librustc_mir/borrow_check/mod.rs#L1637-L1647

in the normal case, we take that `None` path, and thus the mode becomes `LocalMutationIsAllowed::Yes`. Then we recurse here:

https://github.com/rust-lang/rust/blob/75af15ee6ca0c12b699a17984b033363cd25e9c3/src/librustc_mir/borrow_check/mod.rs#L1649

The key point then, in terms of your PR, is that we **started** with `LocalMutationIsAllowed::No`, but by the time we reached the local variable, we had switched `Yes`. So in your code it gets back a `Ok(Local)` and checks the flag and sees `No` (the initial value), so we record the mut as used, but that's wrong.

What we could do -- though I don't love it, but it's probably the right thing for now -- is to return in the `Ok` case not just the `Place` but *also* the flag. Then when we decide whether to modify the `used_muts` set, we can check the flag that gets returned rather than the initial value.

File: src/librustc_mir/borrow_check/mod.rs [hunk 4]
```diff
@@ -266,6 +267,11 @@ fn do_mir_borrowck<'a, 'gcx, 'tcx>(
         if let ClearCrossCrate::Set(ref vsi) = mbcx.mir.visibility_scope_info {
             let local_decl = &mbcx.mir.local_decls[local];
 
+            // Skip closures in fn arg positions
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: Nit: can you rephrase as "skip implicit `self` argument for closures"? I'm not sure I would understand the existing comment =)

**Comment by KiChjang** [comment]
We've passed the `lint-unused-mut-variables.rs` compile fail test, but in return `issue-16671.rs` is throwing us errors:

```rust
fn foo<F: FnOnce()>(_f: F) { }

fn main() {
    let mut var = Vec::new();
    foo(move|| {
        var.push(1);
    });
}
```

This looks very much like that the code doesn't understand that `var.push(1);` is a mutable use of the `var` variable, and so it is justified to have a `mut` annotation. The MIR for the closure looks like this:

```rust
fn main::{{closure}}(_1: [closure@.\src\test\run-pass\issue-16671.rs:19:9: 21:6 var:std::vec::Vec<i32>]) -> (){
    let mut _0: ();                      // return place
    let mut _2: ();
    let mut _3: &mut std::vec::Vec<i32>;

    bb0: {                              
        StorageLive(_3);                 // bb0[0]: scope 0 at .\src\test\run-pass\issue-16671.rs:20:9: 20:12
        _3 = &mut (_1.0: std::vec::Vec<i32>); // bb0[1]: scope 0 at .\src\test\run-pass\issue-16671.rs:20:9: 20:12
        _2 = const <std::vec::Vec<T>>::push(move _3, const 1i32) -> [return: bb2, unwind: bb3]; // bb0[2]: scope 0 at .\src\test\run-pass\issue-16671.rs:20:9: 20:20
                                         // ty::Const
                                         // + ty: for<'r> fn(&'r mut std::vec::Vec<i32>, i32) {<std::vec::Vec<T>><i32>::push}
                                         // + val: Value(ByVal(Undef))
                                         // mir::Constant
                                         // + span: .\src\test\run-pass\issue-16671.rs:20:9: 20:20
                                         // + ty: for<'r> fn(&'r mut std::vec::Vec<i32>, i32) {<std::vec::Vec<T>><i32>::push}
                                         // + literal: const <std::vec::Vec<T>>::push
                                         // ty::Const
                                         // + ty: i32
                                         // + val: Value(ByVal(Bytes(1)))
                                         // mir::Constant
                                         // + span: .\src\test\run-pass\issue-16671.rs:20:18: 20:19
                                         // + ty: i32
                                         // + literal: const 1i32
    }

    bb1: {                               // cleanup
        resume;                          // bb1[0]: scope 0 at .\src\test\run-pass\issue-16671.rs:19:9: 21:6
    }

    bb2: {                              
        StorageDead(_3);                 // bb2[0]: scope 0 at .\src\test\run-pass\issue-16671.rs:20:19: 20:20
        _0 = ();                         // bb2[1]: scope 0 at .\src\test\run-pass\issue-16671.rs:19:16: 21:6
        drop(_1) -> [return: bb4, unwind: bb1]; // bb2[2]: scope 0 at .\src\test\run-pass\issue-16671.rs:21:5: 21:6
    }

    bb3: {                               // cleanup
        drop(_1) -> bb1;                 // bb3[0]: scope 0 at .\src\test\run-pass\issue-16671.rs:21:5: 21:6
    }

    bb4: {                              
        return;                          // bb4[0]: scope 0 at .\src\test\run-pass\issue-16671.rs:21:6: 21:6
    }
}
```

So, it looks like that `_3 = &mut (_1.0: std::vec::Vec<i32>);` is not considered a legitimate mutable use of `_1.0` (which is the upvar `vec`), and neither is `_2 = const <std::vec::Vec<T>>::push(move _3, const 1i32) -> [return: bb2, unwind: bb3];`. I'm thinking that I may need to check the Rvalues as well, and see if it takes a mutable reference of a variable, and if so, mark that variable as a `used_mut`. I don't feel like this is the correct fix; @nikomatsakis do you have any better suggestions?

File: src/librustc_mir/borrow_check/mod.rs [hunk 5]
```diff
@@ -1524,47 +1598,89 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
-
-                    let mut err = self.tcx.cannot_assign(span, &item_msg, Origin::Mir);
-                    err.span_label(span, "cannot mutate");
-
-                    if place != place_err {
-                        if let Some(name) = self.describe_place(place_err) {
-                            err.note(&format!("Value not mutable causing this error: `{}`", name));
+                match self.is_mutable(place, is_local_mutation_allowed) {
+                    Ok((Place::Local(local), mut_allowed)) => {
+                        if mut_allowed != LocalMutationIsAllowed::Yes {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: feels like we want a helper function -- this code is the same as above, right?

File: src/librustc_mir/borrow_check/mod.rs [hunk 6]
```diff
@@ -1600,25 +1716,28 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         &self,
         place: &'d Place<'tcx>,
         is_local_mutation_allowed: LocalMutationIsAllowed,
-    ) -> Result<(), &'d Place<'tcx>> {
+    ) -> Result<(&'d Place<'tcx>, LocalMutationIsAllowed), &'d Place<'tcx>> {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/mod.rs]: Nit: might be worth making a struct here for this tuple, so we can give names to the fields and write comments explaining what is going on.

**Comment by nikomatsakis** [comment]
@KiChjang 

OK, I've been looking at this a bit more and I have some thoughts on how to fix. I think the problem is "rooted" in this match:

https://github.com/rust-lang/rust/blob/2253216ec500601f9c2998196cb1b0a1ed0ae576/src/librustc_mir/borrow_check/mod.rs#L1685-L1694

(That link is from master, but the code is the same in your branch.)

In particular, in the [broken case](https://github.com/rust-lang/rust/pull/48605#issuecomment-376046835) here, we will I believe go down this line:

https://github.com/rust-lang/rust/blob/2253216ec500601f9c2998196cb1b0a1ed0ae576/src/librustc_mir/borrow_check/mod.rs#L1691

This indicates that the variable was declared as mutable in the surrounding context. However, in that event, we currently recurse. Given that we [already know this is an upvar field projection](https://github.com/rust-lang/rust/blob/2253216ec500601f9c2998196cb1b0a1ed0ae576/src/librustc_mir/borrow_check/mod.rs#L1679), this recursion *may* be necessary -- e.g., we want to ensure that this test fails:

```
fn foo<T: Fn()>(t: T) { t() }

fn bar() {
  let mut x = 3;
  foo(|| x += 1);
}
```

Here, the closure will be inferred to `Fn`, and hence have only `&self`, so the mutability check should fail (however, we may get errors earlier on). Presuming the recursion does not yield an error, though, we probably ought to be overriding the return value so that it returns this outer value. Something like:

```rust
                            match (decl.mutability, is_local_mutation_allowed) {
                                (Mutability::Not, LocalMutationIsAllowed::No)
                                | (Mutability::Not, LocalMutationIsAllowed::ExceptUpvars) => {
                                    Err(place)
                                }
                                (Mutability::Not, LocalMutationIsAllowed::Yes)
                                | (Mutability::Mut, _) => {
                                    let _ = self.is_mutable(&proj.base, is_local_mutation_allowed)?;
                                    Ok((place, is_local_mutation_allowed))
                                }
                            }
```

This should then cause the field to be added to `used_mut_upvars`

**Comment by nikomatsakis** [comment]
OK @KiChjang, just to show you what I mean, I pushed a commit -- it doesn't fix the problem though. I didn't dig deeper to find out why. I still think this is the right track, though I could be confused.

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/363195729). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (605048/605048), completed with 4772 local objects.
---
[00:00:46] configure: rust.quiet-tests     := True
---
[01:01:38] ..........................................................................i.........................
[01:01:46] .................i..................................................................................
---
[01:02:29] ..............................................................................................i.....
[01:02:38] ....................................................................i...............................
---
[01:03:53] .............................................i......................................................
---
[01:07:43] .....................................................F..............................................
[01:08:02] ....................................................................................................
[01:08:23] ....................................................................................................
[01:08:52] ....................................................................................................
[01:09:16] ..............................i.....................................................................
[01:09:34] ...............................................................i....................................
[01:09:54] ................................................i...................................................
[01:10:21] ....................................................................................................
[01:10:49] ....................................................................................................
[01:11:18] ....................................................................................................
[01:11:51] ...i...............................................................................................i
[01:12:08] .......................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[01:12:40] .............................................................
[01:13:20] ....................................................................................................
[01:14:09] ...............................................................ii...................................
[01:14:47] ..........................i......................test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[01:15:13] ..............................i.ii.................
[01:16:08] .......................................................................................iiiiiii......
---
[01:19:01] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/issue-16671.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issue-16671.stage2-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Zmiri" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "borrowck=compare" "-Z" "emit-end-regions" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issue-16671.stage2-x86_64-unknown-linux-gnu.aux"
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

**Comment by KiChjang** [comment]
@TimNN The log above doesn't show the relevant snippet for the reason why the PR failed travis.

**Comment by nikomatsakis** [comment]
Current status: @KiChjang is going to investigate a bit further in the direction hinted at by my last comment (or maybe me, if I get some extra time). Marking as S-waiting-on-author.

File: src/librustc_mir/borrow_check/mod.rs [hunk 7]
```diff
@@ -1524,47 +1598,89 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                            // mutated, then it is justified to be annotated with the `mut`
+                            // keyword, since the mutation may be a possible reassignment.
+                            let mpi = self.move_data.rev_lookup.find_local(*local);
+                            if flow_state.inits.contains(&mpi) {
+                                self.used_mut.insert(*local);
+                            }
+                        }
+                    }
+                    Ok((Place::Projection(ref proj), _mut_allowed)) => {
+                        if let Some(field) = self.is_upvar_field_projection(&proj.base) {
```
Comment (KiChjang) [inline src/librustc_mir/borrow_check/mod.rs]: So, the code doesn't work because of this line -- it's checking whether the _base projection_ is an upvar field projection, which clearly isn't the case, so instead we need to check the projection itself. I don't know if this would change anything else, but I'm going to try making this small change to see what results.

**Comment by KiChjang** [comment]
We're almost there! The one final piece of the puzzle stems from the following basic block:

```rust
    bb2: {                              
        StorageLive(_3);                 // bb2[0]: scope 1 at .\src\test\run-pass\issue-16671.rs:19:9: 21:6
        StorageLive(_4);                 // bb2[1]: scope 1 at .\src\test\run-pass\issue-16671.rs:19:9: 21:6
        _4 = move _1;                    // bb2[2]: scope 1 at .\src\test\run-pass\issue-16671.rs:19:9: 21:6
        _3 = [closure@.\src\test\run-pass\issue-16671.rs:19:9: 21:6] { var: move _4 }; // bb2[3]: scope 1 at .\src\test\run-pass\issue-16671.rs:19:9: 21:6
                                         // closure
                                         // + def_id: DefId(0/1:10 ~ issue_16671[317d]::main[0]::{{closure}}[0])
                                         // + substs: ClosureSubsts {
                                         //     substs: Slice(
                                         //         [
                                         //             i32,
                                         //             extern "rust-call" fn(()),
                                         //             std::vec::Vec<i32>
                                         //         ]
                                         //     )
                                         // }
        drop(_4) -> [return: bb5, unwind: bb4]; // bb2[4]: scope 1 at .\src\test\run-pass\issue-16671.rs:21:5: 21:6
    }
```

Now, the code is able to correctly flag `_4` as being a used mutable variable; however `_4` is _not_ a user-defined variable. The real user-defined variable is `_1`, as seen in this MIR statement:

```rust
    _4 = move _1;
```
So I'm currently thinking if there's a way to track a variable's "path of assignments" throughout a basic block, i.e. given `_4` here, I'd like to have an array of `[_1, _4]`, so we can properly mark this entire array of locals as used muts. @nikomatsakis @pnkfelix do you have any suggestions?

**Comment by KiChjang** [comment]
I don't understand why that UI test is failing; I'm able to reproduce it by running `./x.py test src/test/ui --test-args suggestions`, but when I compile `suggestions.rs` directly, I don't see the unused mut warning duplicated at all. What's going on here?

**Review by nikomatsakis — APPROVED** [review]
This is awesome work. 💯

**Comment by kennytm** [comment]
@bors retry

Assuming the 30-minute timeout is spurious.
